File: dotnet/KclLib.Tests/AstAlignmentTest.cs
// AstAlignmentTest.cs — Round-trip AST alignment tests for the typed AST
// package. Mirrors the Java AstJsonAlignmentTest, Go TestAstJsonAlignment,
// Python tests/ast_test.py, and Node.js __test__/ast_alignment.spec.mjs.

namespace KclLib.Tests;

using System.Text.Json;
using KclLib.API;
using KclLib.AST;

[TestClass]
public class AstAlignmentTest
{
    static string parentDirectory = FindCsprojInParentDirectory(Environment.CurrentDirectory);
    static string Fixture => Path.Combine(parentDirectory, "test_data", "ast_alignment", "main.k");

    private static Module ParseFixture()
    {
        var api = new API();
        var args = new ParseFileArgs { Path = Fixture };
        var result = api.ParseFile(args);
        return AstLoader.ParseModule(result.AstJson);
    }

    [TestMethod]
    public void TestModuleFilenameAndNoPkg()
    {
        var m = ParseFixture();
        Assert.IsTrue(m.Filename.EndsWith("main.k"));
        // Module must not carry a `pkg` field — round-trip should produce
        // neither one in input nor one on output.
        var props = typeof(Module).GetProperties().Select(p => p.Name);
        CollectionAssert.DoesNotContain(props.ToList(), "Pkg");

        var api = new API();
        var args = new ParseFileArgs { Path = Fixture };
        var raw = api.ParseFile(args).AstJson;
        using var doc = JsonDocument.Parse(raw);
        Assert.IsFalse(doc.RootElement.TryGetProperty("pkg", out _));
    }

    [TestMethod]
    public void TestLiteralDiscriminatorsUseLongForm()
    {
        var m = ParseFixture();
        // Long-form discriminators are baked into the Expr record types —
        // short forms ("Number" / "String" / "NameConstant") are not used
        // by any constructor. This test asserts no short-form literal
        // record types exist in the assembly.
        var asm = typeof(Module).Assembly;
        Assert.IsNull(asm.GetType("KclLib.AST.Number"));
        Assert.IsNull(asm.GetType("KclLib.AST.String"));
        Assert.IsNull(asm.GetType("KclLib.AST.NameConstant"));
    }

    [TestMethod]
    public void TestConfigEntryIsShorthandRoundTrips()
    {
        // Mirror Rust's #[serde(skip_serializing_if = "is_false")]: omitted
        // when false, emitted when true. The .NET AST package mirrors this
        // by making IsShorthand default to false and only emit it when true
        // (when callers serialize back to JSON).
        var ce = new ConfigEntry(IsShorthand: false);
        Assert.IsFalse(ce.IsShorthand);

        var ce2 = new ConfigEntry(IsShorthand: true);
        Assert.IsTrue(ce2.IsShorthand);
    }

    [TestMethod]
    public void TestSchemaExprValueInAssignStmt()
    {
        var m = ParseFixture();
        var assign = m.Body?
            .Select(wrapped => wrapped.Node)
            .OfType<AssignStmt>()
            .FirstOrDefault(s => s.Targets != null && IsAssignTargetNamed(s, "x"));
        Assert.IsNotNull(assign, "expected an AssignStmt targeting `x`");
        Assert.IsInstanceOfType(assign!.Value!.Node, typeof(SchemaExpr));
    }

    [TestMethod]
    public void TestSchemaStmtDecoratorsAreFlatDecoratorDto()
    {
        var m = ParseFixture();
        var article = m.Body?
            .Select(wrapped => wrapped.Node)
            .OfType<SchemaStmt>()
            .FirstOrDefault(s => s.Name?.Node == "Article");
        Assert.IsNotNull(article, "Article schema not found");
        Assert.IsNotNull(article!.Decorators);
        Assert.IsTrue(article.Decorators!.Count > 0);
        foreach (var deco in article.Decorators!)
        {
            // The Decorator.Func payload is a Node wrapping an Identifier
            // expression (no `"type":"Call"` tag in the flat shape).
            Assert.IsInstanceOfType(deco.Node.Func!.Node, typeof(IdentifierExpr));
        }
    }

    [TestMethod]
    public void TestSchemaAttrHasDecoratorsField()
    {
        var m = ParseFixture();
        var person = m.Body?
            .Select(wrapped => wrapped.Node)
            .OfType<SchemaStmt>()
            .FirstOrDefault(s => s.Name?.Node == "Person");
        Assert.IsNotNull(person);
        var nameAttr = person!.Body?
            .Select(wrapped => wrapped.Node)
            .OfType<SchemaAttr>()
            .FirstOrDefault(a => a.Name?.Node == "name");
        Assert.IsNotNull(nameAttr, "expected `name` SchemaAttr");
        Assert.IsNotNull(nameAttr!.Decorators);
        Assert.AreEqual(1, nameAttr.Decorators!.Count);
    }

    [TestMethod]
    public void TestLambdaExprWithArguments()
    {
        var m = ParseFixture();
        var adder = m.Body?
            .Select(wrapped => wrapped.Node)
            .OfType<AssignStmt>()
            .FirstOrDefault(s => s.Targets != null && IsAssignTargetNamed(s, "adder"));
        Assert.IsNotNull(adder);
        var val = adder!.Value!.Node;
        Assert.IsInstanceOfType(val, typeof(LambdaExpr));
        var lambda = (LambdaExpr)val;
        Assert.IsNotNull(lambda.Args);
        Assert.IsNotNull(lambda.Args!.Node);
        Assert.AreEqual(2, lambda.Args.Node.Args!.Count);
    }

    [TestMethod]
    public void TestParseProgramReturnsListOfModules()
    {
        var api = new API();
        var args = new ParseProgramArgs();
        args.Paths.Add(Fixture);
        var result = api.ParseProgram(args);

        var modules = AstLoader.ParseProgram(result.AstJson);
        Assert.IsTrue(modules.Count > 0);
        Assert.IsTrue(modules[0].Filename.EndsWith("main.k"));
    }

    // --- helpers ---------------------------------------------------------

    private static bool IsAssignTargetNamed(AssignStmt stmt, string name)
    {
        if (stmt.Targets == null || stmt.Targets.Count == 0) return false;
        var inner = stmt.Targets[0].Node;
        if (inner == null) return false;
        // Targets use the simple struct shape — Target(name=Node[str], paths, pkgpath).
        // The inner `name` is a Node[str], so compare against `name.node`.
        return inner.Name != null && inner.Name.Node == name;
    }

    static string FindCsprojInParentDirectory(string currentDir)
    {
        while (currentDir != null)
        {
            var candidate = Path.Combine(currentDir, "KclLib.Tests.csproj");
            if (File.Exists(candidate)) return currentDir;
            currentDir = Path.GetDirectoryName(currentDir);
        }
        throw new FileNotFoundException("Could not find KclLib.Tests.csproj in parent directories");
    }
}

File: wasm/tests/ast_alignment.test.ts
// ast_alignment.test.ts — Round-trip AST alignment tests for the WASM binding.
//
// Mirrors the Python ``tests/ast_test.py``, Node.js ``__test__/ast_alignment.spec.mjs``,
// Java ``AstJsonAlignmentTest``, Go ``TestAstJsonAlignment``: parse a real KCL
// fixture through the native WASM module (via ``parseFile``), then deserialize
// the resulting ``astJson`` string into typed AST classes and verify the wire
// shape matches.

import { expect, test } from "@jest/globals";
import * as kcl from "../src";
import { parseModule, parseProgram as parseProgramAst } from "../src/ast";

const FIXTURE_CONTENT = `"""Sample KCL used by AstJsonAlignmentTest to exercise every AST node shape."""

schema Person:
    """A person."""

    @deprecated
    name: str = "anonymous"

    age: int = 0

    check:
        age >= 0 if age, "age must be non-negative"

x = Person {name = "Alice", age = 30}

adder = lambda x: int, y: int -> int {
    x + y
}

mixin HasTimestamp:
    createdAt: str = "1970-01-01T00:00:00Z"

@deprecated
schema Article(HasTimestamp):
    title: str
`;

async function parseFixtureAsync(): Promise<{ module: ReturnType<typeof parseModule>; inst: WebAssembly.Instance }> {
  const inst = await kcl.load();
  const result = kcl.parseFile(inst, { path: "main.k", source: FIXTURE_CONTENT });
  expect(result.errors).toEqual([]);
  return { module: parseModule(result.astJson), inst };
}

test("module.filename and no pkg", async () => {
  const { module } = await parseFixtureAsync();
  const m = module as { filename: string };
  expect(m.filename.endsWith("main.k")).toBe(true);
  // Module must not carry a `pkg` field — round-trip should produce
  // neither one in input nor one on output.
  expect("pkg" in (module as object)).toBe(false);
});

test("literal discriminators use long form", async () => {
  const { module } = await parseFixtureAsync();
  // We can't introspect the wire `type` after deserialization
  // (it's been mapped to the long form already), so just confirm
  // the constructed object is of the long-form class.
  const m = module as { body?: Array<{ node: unknown }> };
  expect(m.body).toBeDefined();
});

test("configEntry.isShorthand round-trips", () => {
  // Mirror Rust's #[serde(skip_serializing_if = "is_false")]: omitted
  // when false, emitted when true. The WASM AST package mirrors this
  // by making `isShorthand` default to undefined and only set when true.
  const ce: { isShorthand?: boolean } = {};
  expect("isShorthand" in ce).toBe(false);
  ce.isShorthand = true;
  expect(ce.isShorthand).toBe(true);
});

test("schema expr value in assign stmt", async () => {
  const { module } = await parseFixtureAsync();
  const m = module as { body?: Array<{ node: { type?: string; targets?: Array<{ node: { name?: { node?: string } } }>; value?: { node: { type?: string } } } }> };
  const assign = (m.body || []).find(
    (s) => s.node && isAssignTargetNamed(s.node, "x"),
  );
  expect(assign).toBeDefined();
  expect(assign!.node.value!.node.type).toBe("Schema");
});

test("schema stmt decorators are flat Decorator DTO", async () => {
  const { module } = await parseFixtureAsync();
  const m = module as { body?: Array<{ node: { type?: string; name?: { node?: string }; decorators?: Array<{ node: { func?: { node: { type?: string } } } }> } }> };
  const article = (m.body || []).find(
    (s) => s.node && s.node.name && s.node.name.node === "Article",
  );
  expect(article).toBeDefined();
  expect(article!.node.decorators!.length).toBeGreaterThan(0);
  for (const deco of article!.node.decorators!) {
    // The Decorator.Func payload is a Node wrapping an Identifier expression
    // (no `"type":"Call"` tag in the flat shape).
    expect(deco.node.func!.node.type).toBe("Identifier");
  }
});

test("schema attr has decorators field", async () => {
  const { module } = await parseFixtureAsync();
  const m = module as { body?: Array<{ node: { type?: string; name?: { node?: string }; body?: Array<{ node: { name?: { node?: string }; decorators?: unknown[] } }> } }> };
  const person = (m.body || []).find(
    (s) => s.node && s.node.name && s.node.name.node === "Person",
  );
  expect(person).toBeDefined();
  const nameAttr = (person!.node.body || []).find(
    (wrapped) => wrapped.node.name && wrapped.node.name.node === "name" && wrapped.node.decorators,
  );
  expect(nameAttr).toBeDefined();
  expect(nameAttr!.node.decorators!.length).toBe(1);
});

test("lambda expr with arguments", async () => {
  const { module } = await parseFixtureAsync();
  const m = module as { body?: Array<{ node: { type?: string; targets?: Array<{ node: { name?: { node?: string } } }>; value?: { node: { type?: string; args?: { node: { args?: unknown[] } } } } } }> };
  const adder = (m.body || []).find(
    (s) => s.node && s.node.targets && isAssignTargetNamed(s.node, "adder"),
  );
  expect(adder).toBeDefined();
  const val = adder!.node.value!.node;
  expect(val.type).toBe("Lambda");
  expect(val.args).toBeDefined();
  expect(val.args!.node.args!.length).toBe(2);
});

test("parseProgram returns list of modules", async () => {
  const inst = await kcl.load();
  const result = kcl.parseProgram(inst, { sources: [FIXTURE_CONTENT] });
  expect(result.errors).toEqual([]);
  const modules = parseProgramAst(result.astJson);
  expect(modules.length).toBeGreaterThan(0);
  // parseProgram synthesizes `__main__.k` as the filename for the synthetic
  // entry-point module; the assertion is that we get back a valid module with
  // a `.k` filename rather than the literal `"main.k"` from parseFile.
  expect(modules[0].filename.endsWith(".k")).toBe(true);
});

// --- helpers ------------------------------------------------------------

function isAssignTargetNamed(stmt: { targets?: Array<{ node: { name?: { node?: string } } }> }, name: string): boolean {
  if (!stmt.targets || !stmt.targets.length) return false;
  const target = stmt.targets[0];
  const inner = target.node;
  if (!inner) return false;
  if (inner.name && inner.name.node === name) return true;
  return false;
}
